src/Components/Components/src/NavigationManagerExtensions.cs

```diff
@@ -0,0 +1,694 @@
+        };
+
+        private static string Format(bool value)
+            => value.ToString(CultureInfo.InvariantCulture);
+
+        private static string? Format(bool? value)
+            => value?.ToString(CultureInfo.InvariantCulture);
+
+        private static string Format(DateTime value)
+            => value.ToString(CultureInfo.InvariantCulture);
+
+        private static string? Format(DateTime? value)
+            => value?.ToString(CultureInfo.InvariantCulture);
+
+        private static string Format(decimal value)
+            => value.ToString(CultureInfo.InvariantCulture);
+
+        private static string? Format(decimal? value)
+            => value?.ToString(CultureInfo.InvariantCulture);
+
+        private static string Format(double value)
+            => value.ToString(CultureInfo.InvariantCulture);
+
+        private static string? Format(double? value)
+            => value?.ToString(CultureInfo.InvariantCulture);
+
+        private static string Format(float value)
+            => value.ToString(CultureInfo.InvariantCulture);
+
+        private static string? Format(float? value)
+            => value?.ToString(CultureInfo.InvariantCulture);
+
+        private static string Format(Guid value)
+            => value.ToString(null, CultureInfo.InvariantCulture);
+
+        private static string? Format(Guid? value)
+            => value?.ToString(null, CultureInfo.InvariantCulture);
+
+        private static string Format(int value)
+            => value.ToString(CultureInfo.InvariantCulture);
+
+        private static string? Format(int? value)
+            => value?.ToString(CultureInfo.InvariantCulture);
+
+        private static string Format(long value)
+            => value.ToString(CultureInfo.InvariantCulture);
+
+        private static string? Format(long? value)
+            => value?.ToString(CultureInfo.InvariantCulture);
+
+        private struct QueryStringBuilder
+        {
+            private readonly StringBuilder _builder;
+
+            private bool _hasNewParameters;
+
+            public string UriWithQueryString => _builder.ToString();
+
+            public QueryStringBuilder(ReadOnlySpan<char> uriWithoutQueryString)
+            {
+                _builder = new();
+                _builder.Append(uriWithoutQueryString);
+
+                _hasNewParameters = false;
+            }
+
+            public void AppendParameter(ReadOnlySpan<char> encodedName, ReadOnlySpan<char> encodedValue)
+            {
+                if (!_hasNewParameters)
+                {
+                    _hasNewParameters = true;
+                    _builder.Append('?');
+                }
+                else
+                {
+                    _builder.Append('&');
+                }
+
+                _builder.Append(encodedName);
+                _builder.Append('=');
+                _builder.Append(encodedValue);
+            }
+        }
+
+        private class ParameterData
```

Comment:
Looks like this could be a `struct` if we wanted. It could either be:

 * Mutable, so you can set its `DidReplace` after creation. It's OK for structs to be mutable as long as they aren't part of the public API and they are only used by a small enough chunk of our code that we can reason about it.
 * Readonly, and then you'd have some constructor that makes a new copy with a different `DidReplace` flag to overwrite the previous dictionary entry when it needs to change

It's not a big deal but would save one allocation per dictionary entry.

Reply:
I initially tried to do the mutable struct route, but found there wasn't a way to re-obtain and mutate the struct without making a copy (looking up the value in the dictionary returns a copy, unless maybe there's a way to achieve this that I'm not aware of). In which case, we would have to mutate the copy and re-insert it into the dictionary, and I wasn't sure if the cost associated with that would be worth using a `struct`. Maybe even if it is a small speed hit, it could be worth it to minimize GC pressure?

If we do go the readonly `struct` route, maybe a `record struct` could be a way to go? Then we could use `with` to make the copy instead of introducing our own mechanism?

Reply:
You're certainly correct about the copying.

If it's a small struct (and in this case it is - just a string and a bool), then copying it is extremely cheap. It would definitely be fine to copy out of the dictionary and then overwrite the dictionary entry with a different value, and generally preferable to mutating a heap entry that could be anywhere in memory.